memory_unsafe.go

```diff
@@ -0,0 +1,256 @@
+	// space until frozen. As a security precaution, the kernel doesn't allow
+	// mapping bpf map memory as read-write into user space if the bpf map was
+	// frozen, or if it was created using the RDONLY_PROG flag.
+	//
+	// The user would be able to write to the map after freezing (since the kernel
+	// can't change the protection mode of an already-mapped page), while the
+	// verifier assumes the contents to be immutable.
+	//
+	// Map the bpf map memory over a page-aligned allocation on the Go heap.
+	err = mapmap(fd, alloc, size, unix.PROT_READ|unix.PROT_WRITE)
+
+	// If the map is frozen when an rw mapping is requested, expect EPERM. If the
+	// map was created with BPF_F_RDONLY_PROG, expect EACCES.
+	var ro bool
+	if errors.Is(err, unix.EPERM) || errors.Is(err, unix.EACCES) {
+		ro = true
+		err = mapmap(fd, alloc, size, unix.PROT_READ)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("setting up memory-mapped region: %w", err)
+	}
+
+	mm := &Memory{
+		unsafe.Slice((*byte)(alloc), size),
+		ro,
+		true,
+	}
+
+	return mm, nil
+}
+
+// allocate returns a pointer to a page-aligned section of memory on the Go
+// heap, managed by the runtime.
+func allocate(size int) (unsafe.Pointer, error) {
+	// Memory-mapping over a piece of the Go heap is unsafe when the GC can
+	// randomly decide to move objects around, in which case the mapped region
+	// will not move along with it.
+	if heapObjectsCanMove() {
+		return nil, errors.New("this Go runtime has a moving garbage collector")
+	}
+
+	if size == 0 {
+		return nil, errors.New("size must be greater than 0")
+	}
+
+	// Request at least two pages of memory from the runtime to ensure we can
+	// align the requested allocation to a page boundary. This is needed for
+	// MAP_FIXED and makes sure we don't mmap over some other allocation on the Go
+	// heap.
+	size = internal.Align(size+os.Getpagesize(), os.Getpagesize())
+
+	// Allocate a new slice and store a pointer to its backing array.
+	alloc := unsafe.Pointer(unsafe.SliceData(make([]byte, size)))
+
+	// Align the pointer to a page boundary within the allocation. This may
+	// alias the initial pointer if it was already page-aligned.
+	aligned := unsafe.Pointer(internal.Align(uintptr(alloc), uintptr(os.Getpagesize())))
+	runtime.KeepAlive(alloc)
+
+	// Return an aligned pointer into the backing array, losing the original
+	// reference. The runtime.SetFinalizer docs specify that its argument 'must be
+	// a pointer to an object, complit or local var', but this is still somewhat
+	// vague and not enforced by the current implementation.
+	//
+	// Currently, finalizers can be set and triggered from any address within a
+	// heap allocation, even individual struct fields or arbitrary offsets within
+	// a slice. In this case, finalizers set on struct fields or slice offsets
+	// will only run when the whole struct or backing array are collected. The
+	// accepted runtime.AddCleanup proposal makes this behaviour more explicit and
```

Comment:
Does all of this still work when switching to AddCleanup?

Reply:
Need to check, but afaik AddCleanup is SetFinalizer modulo object resurrection.